### ISSUE TITLE
Is it possible a Class reading from DB that has length 1 byte + e.g (2 bits)?

### ISSUE THREAD

**Issue by Mallowan** [rapha-dev/examples#2]
@rapha-dev,
 Just have tried to add couple of bits to DB  that had only one byte length.
And I've got an exeption... Of course I've checked the length of both ( in PLC project and VS project).

Perhaps, do I need to use 2 entire Bytes for Class reading?

P.S. If I'm going to read something from newly created DB It goes well. True, it happens always after "rain dance". )
Mallowan

**Comment by rapha-dev** [comment]
You are correct, minimum is 2 bytes on either class or struct reading. It's not limited by S7NetPlus but Siemens contraints any DB to minimum of 2 bytes in length. There is a known leak on reading less than 2 bytes as well, maybe this can be fixed in some time.

**Comment by Mallowan** [comment]
P.S. If I'm going to read something from newly created DB It goes well. True, it happens always after "rain dance". )

**Comment by Mallowan** [comment]
> You are correct, minimum is 2 bytes on either class or struct reading. It's not limited by S7NetPlus but Siemens contraints any DB to minimum of 2 bytes in length. There is a known leak on reading less than 2 bytes as well, maybe this can be fixed in some time.


No, no, no...
Perhaps so: If I need to read by Class method I must have at least 1 entire Byte, or 2 entire Bytes or 3 Bytes.
Always must be "Entrire Byte".

Because I have   a few DBs with length 1 Byte and they are readable excellently.
Just I had braked wholeness (added two boolean addresses) of first Byte an have got an exeption.

Mallowan

**Comment by rapha-dev** [comment]
Are you using latest Release which is 0.1.9?
There was a fix regarding byte count in Bits lately. So reading 8 Bits should be fine, even a single Bit works (not tested in class reading though).
I will check and test reading Bits only.

**Comment by Mallowan** [comment]
No I'm using 0.1.8.
It seems, there's 0.1.8 at this webpage available, as far as I know.

**Comment by rapha-dev** [comment]
There is a newer version since a few days.
Also have tested reading 8 bools via ReadClass, it works w/o Exception. Will update example.

**Comment by Mallowan** [comment]
OK. Is there a list with fixed bugs there? Just to read about...

**Comment by rapha-dev** [comment]
Well, not ATM, we're working on a concept to publish these alongside any Release. You may use GitHub history for now.
And make sure your target framework is at least net452 as any upcoming versions will rely on this.

**Comment by Mallowan** [comment]
I've checked, 0.1.8 only available here.

**0.1.8** 
I 'm reading 8 bools exellently. I've got an exeption after addition of 2 bits to these 8 bools (entire byte).

**Comment by rapha-dev** [comment]
https://github.com/killnine/s7netplus/releases
0.1.9 is available in NuGet, also.

**Comment by rapha-dev** [comment]
Have added another example https://github.com/rapha-dev/examples/tree/master/S7.Net.Examples
It reads 10 Bits via ReadClass, can be easily expanded to more Bits (or other Types).

**Comment by Mallowan** [comment]
Thank you... I'll look at that.
Can't believe, a neighboring DB wiith 40 Bytes lenght is being read without any exception.
Just needed to create a new one with 2 Bytes length. And I've stuck in this smaller one.

**Comment by rapha-dev** [comment]
It is a S7-1500? Have a live PLC here, maybe the Simulator gives false positives (but data is shown correctly).

**Comment by Mallowan** [comment]
Yes, 1511, real hardware...

**Comment by rapha-dev** [comment]
So...reading 10 bits as in `S7.Net.Examples` from a real S7-1500 works in my case. Test-DB looks like this.
![block_remanent_vars](https://user-images.githubusercontent.com/814049/42165713-2e3b42f6-7e09-11e8-8df4-896d7d8beed6.PNG)

@Mallowan 
Have you disabled `Optimized access` in block settings for the newly created block?
![block_not_optimized](https://user-images.githubusercontent.com/814049/42165926-c593ec7a-7e09-11e8-8cc4-c2e7887e9c36.PNG)

**Comment by Mallowan** [comment]
@rapha-dev ,
FYI, 
![fyi_01](https://user-images.githubusercontent.com/37699980/42184947-de7adeba-7e4f-11e8-8d30-ffc1cbe54659.png)
So I would say. It must be created a Class at least equal 1 Byte or 2 Bytes or 3 Bytes. Every time entire byte must be used.

**Comment by rapha-dev** [comment]
@Mallowan 
I doubt it is s7netplus related. Are you willing to show your code?

**Comment by Mallowan** [comment]
Yeahh.. give me a moment. Perhaps I read Class in a weird manner :)

Mallowan

**Comment by rapha-dev** [comment]
Possible that you are stuck in issue [#135 ReadClass and ReadStruct order is not guaranteed](https://github.com/killnine/s7netplus/issues/135).

**Comment by Mallowan** [comment]
@rapha-dev,
All my DBs I read by similar way. Here is 40 bytes DB. Just to load a little  (continuosly) the traffic between nodes while I'm doing other practical things in the project. It reads excellently.


```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace S7net_01.Test_DBs
{
    public class DB1 
    {
        public Double varReal1 { get; set; }
        public int varDInt1 { get; set; }
        public Double varReal2 { get; set; }
        public int varDInt2 { get; set; }
        public Double varReal3 { get; set; }
        public int varDInt3 { get; set; }
        public Double varReal4 { get; set; }
        public int varDInt4 { get; set; }
        public Double varReal5 { get; set; }
        public int varDInt5 { get; set; }
        public Double varReal6 { get; set; }


    }
```
```
   void timer_Tick(object sender, EventArgs e) //Reading by means of System.Timer 
        {


            DB1 db1 = new DB1();
            plc.ReadClass(db1, 1); // reading DB1

            var value0 = db1.varReal1;
            var value1 = db1.varDInt1;
            var value2 = db1.varReal2;
            var value3 = db1.varDInt2;
            var value4 = db1.varReal3;
            var value5 = db1.varDInt3;
            var value6 = db1.varReal4;
            var value7 = db1.varDInt4;
            var value8 = db1.varReal5;
            var value9 = db1.varDInt5;
            var value10 = db1.varReal6;

            label9.Text = value0.ToString(); // variable to label
            label10.Text = value1.ToString(); // variable to label
            label11.Text = value2.ToString(); // variable to label
            label12.Text = value3.ToString(); // variable to label
            label13.Text = value4.ToString(); // variable to label
            label14.Text = value5.ToString(); // variable to label
            label15.Text = value6.ToString(); // variable to label
            label16.Text = value7.ToString(); // variable to label
            label17.Text = value8.ToString(); // variable to label
            label18.Text = value9.ToString();  // variable to label
            label19.Text = value10.ToString(); // variable to label

        }
```
Mallowan

**Comment by rapha-dev** [comment]
@Mallowan 
ReadClass usage seems to be correct, the continously changing data shouldn't be a problem in this case. Data always is read consistent within a single request.

Where is the `ArrayOutOfBoundsException` throwing? I will do some tests on ReadClass with weird configs ;)

**Comment by Mallowan** [comment]
I use try/catch operators for the next DBs reading (others practical things). There was an example above to load traffic, as I said above. To get an exeptions as shown in my post above you need to follow my pic.

Mallowan

**Comment by rapha-dev** [comment]
Sure, and that should work flawlessly.

**Comment by Mallowan** [comment]
I am all for it!. Because this library is base of my experiment trial project!

sent you on google a message.
